test/model.test.js

```diff
@@ -0,0 +1,893 @@
+      });
+  });
+
+  it('should handle options (replace)', function() {
+    var Model = thinky.createModel(util.s8(), {
+      id: String,
+      num: Number
+    });
+
+    return Model.ready()
+      .then(function() {
+        return Model.save({ id: 'foo', bar: 'buzz' });
+      })
+      .then(function(result) {
+        expect(result.id).to.equal('foo');
+        return Model.save({ id: 'foo' }, { conflict: 'replace' });
+      })
+      .then(function(result) {
+        expect(util.deepCopy(result)).to.eql({ id: 'foo' });
+      });
+  });
+});
+
+describe("Joins", function() {
+  var types = [ 'hasOne', 'belongsTo', 'hasMany', 'hasAndBelongsToMany'];
+  types.forEach(function(type) {
+    describe(type, function() {
+      afterEach(function() { return test.dropTables(); });
+
+      it('should throw if it uses a field already used by another relation', function() {
+        var Model = thinky.createModel(util.s8(), { id: String }, { init: false });
+        var OtherModel = thinky.createModel(util.s8(), {
+          id: String,
+          otherId: String
+        }, {
+          init: false
+        });
+
+        var AnotherModel = thinky.createModel(util.s8(), {
+          id: String,
+          otherId: String
+        }, {
+          init: false
+        });
+
+        Model[type](OtherModel, "otherDoc", "id", "otherId", { init: false });
+
+        return Promise.all([ Model.ready(), OtherModel.ready(), AnotherModel.ready() ])
+          .then(function() {
+            expect(function() {
+              Model[type](AnotherModel, "otherDoc", "id", "otherId");
+            }).to.throw("The field `otherDoc` is already used by another relation.");
+          });
+      });
+
+
+      it('belongsTo should throw if the first argument is not a model', function() {
+        var Model = thinky.createModel(util.s8(), { id: String }, { init: false });
+
+        return Model.ready()
+          .then(function() {
+            expect(function() {
+              Model[type](function() {}, "otherDoc", "otherId", "id");
+            }).to.throw('First argument of `' + type + '` must be a Model');
+          });
+      });
+
+      it('should create an index on the other model', function() {
+        var Model = thinky.createModel(util.s8(), {
+          id: String,
+          foreignKeyName: String
+        });
+
+        var foreignKey = util.s8();
+        var schema = { id: String };
+        schema[foreignKey] = String;
+        var OtherModel = thinky.createModel(util.s8(), schema);
+
+        Model[type](OtherModel, "otherDoc", "modelId", foreignKey);
+        return Promise.all([ Model.ready(), OtherModel.ready() ])
+          .then(function() {
+            return Promise.all([
+              test.r.table(OtherModel.getTableName()).indexList().run(),
+              test.r.table(OtherModel.getTableName()).indexWait(foreignKey).run()
+            ]);
+          });
+      });
+
```

Comment:
For `hasAndBelongsToMany`, there's a test to check for an index on the Model that called `hasAndBelongsToMany`.